data/srv/epages/eproot/Perl/man/man3/cpanfile.3
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "cpanfile 3"
.TH cpanfile 3 "2016-02-12" "perl v5.12.3" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
cpanfile \- A format for describing CPAN dependencies for Perl applications
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\&  requires \*(AqPlack\*(Aq, \*(Aq1.0\*(Aq; # 1.0 or newer
\&  requires \*(AqJSON\*(Aq, \*(Aq>= 2.00, < 2.80\*(Aq;
\&
\&  recommends \*(AqJSON::XS\*(Aq, \*(Aq2.0\*(Aq;
\&  conflicts \*(AqJSON\*(Aq, \*(Aq< 1.0\*(Aq;
\&
\&  on \*(Aqtest\*(Aq => sub {
\&    requires \*(AqTest::More\*(Aq, \*(Aq>= 0.96, < 2.0\*(Aq;
\&    recommends \*(AqTest::TCP\*(Aq, \*(Aq1.12\*(Aq;
\&  };
\&
\&  on \*(Aqdevelop\*(Aq => sub {
\&    recommends \*(AqDevel::NYTProf\*(Aq;
\&  };
\&
\&  feature \*(Aqsqlite\*(Aq, \*(AqSQLite support\*(Aq => sub {
\&    recommends \*(AqDBD::SQLite\*(Aq;
\&  };
.Ve
.SH "VERSION"
.IX Header "VERSION"
This document describes cpanfile format version 1.0.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\f(CW\*(C`cpanfile\*(C'\fR describes \s-1CPAN\s0 dependencies required to execute associated
Perl code.
.SH "SYNTAX"
.IX Header "SYNTAX"
.IP "requires, recommends, suggests, conflicts" 4
.IX Item "requires, recommends, suggests, conflicts"
.Vb 1
\&  requires $module, $version_requirement;
.Ve
.Sp
Describes the requirement for a module. See CPAN::Meta::Spec for
the meanings of each requirement type.
.Sp
When version requirement is omitted, it is assumed that \f(CW0\fR is
passed, meaning any version of the module would satisfy the
requirement.
.Sp
Version requirement can either be a version number or a string that
satisfies \*(L"Version Ranges\*(R" in CPAN::Meta::Spec, such as \f(CW\*(C`>= 1.0, !=
1.1\*(C'\fR.
.Sp
Note that, per CPAN::Meta::Spec, when a plain version number is
given, it means the version \fIor newer\fR is required. If you want a
specific version for a module, use the specific range syntax, i.e.
\&\f(CW\*(C` == 2.1 \*(C'\fR.
.IP "on" 4
.IX Item "on"
.Vb 1
\&  on $phase => sub { ... };
.Ve
.Sp
Describe requirements for a specific phase. Available phases are
\&\f(CW\*(C`configure\*(C'\fR, \f(CW\*(C`build\*(C'\fR, \f(CW\*(C`test\*(C'\fR, \f(CW\*(C`runtime\*(C'\fR and \f(CW\*(C`develop\*(C'\fR.
.IP "feature" 4
.IX Item "feature"
.Vb 1
\&  feature $identifier, $description => sub { ... };
.Ve
.Sp
Group requirements with features. Description can be omitted, when it
is assumed to be the same as identifier. See
\&\*(L"optional_features\*(R" in CPAN::Meta::Spec for more details.
.IP "configure_requires, build_requires, test_requires, author_requires" 4
.IX Item "configure_requires, build_requires, test_requires, author_requires"
.Vb 2
\&  configure_requires $module, $version;
\&  # on \*(Aqconfigure\*(Aq => sub { requires $module, $version }
\&
\&  build_requires $module, $version;
\&  # on \*(Aqbuild\*(Aq => sub { requires $module, $version };
\&
\&  test_requires $module, $version;
\&  # on \*(Aqtest\*(Aq => sub { requires $module, $version };
\&
\&  author_requires $module, $version;
\&  # on \*(Aqdevelop\*(Aq => sub { requires $module, $version };
.Ve
.Sp
Shortcut for \f(CW\*(C`requires\*(C'\fR in specific phase. This is mainly provided
for compatibilities with Module::Install \s-1DSL\s0.
.SH "USAGE"
.IX Header "USAGE"
\&\f(CW\*(C`cpanfile\*(C'\fR is a format to describe dependencies. How to use this file
is dependent on the tools reading/writing it.
.PP
Usually, you're expected to place the \f(CW\*(C`cpanfile\*(C'\fR in the root of the
directory containing the associated code.
.PP
Tools supporting \f(CW\*(C`cpanfile\*(C'\fR format (e.g. cpanm and carton) will
automatically detect the file and install dependencies for the code to
run.
.PP
There are also tools to support converting cpanfile to \s-1CPAN\s0 toolchain
compatible formats, such as Module::CPANfile,
Dist::Zilla::Plugin::Prereqs::FromCPANfile,
Module::Install::CPANfile, so that \f(CW\*(C`cpanfile\*(C'\fR can be used to
describe dependencies for a \s-1CPAN\s0 distribution as well.
.PP
The cpanfile-dump tool can be used to dump dependencies.
.SH "AUTHOR"
.IX Header "AUTHOR"
Tatsuhiko Miyagawa
.SH "ACKNOWLEDGEMENTS"
.IX Header "ACKNOWLEDGEMENTS"
The format (\s-1DSL\s0 syntax) is inspired by Module::Install and
Module::Build::Functions.
.PP
\&\f(CW\*(C`cpanfile\*(C'\fR specification (this document) is based on Ruby's
Gemfile <http://bundler.io/v1.3/man/gemfile.5.html> specification.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
CPAN::Meta::Spec Module::Install Carton


data/srv/epages/eproot/Perl/lib/site_perl/ExtUtils/MakeMaker/CPANfile.pm
package ExtUtils::MakeMaker::CPANfile;

use strict;
use warnings;
use ExtUtils::MakeMaker ();
use File::Spec::Functions qw/catfile rel2abs/;
use Module::CPANfile;
use version;

our $VERSION = "0.07";

sub import {
  my $class = shift;
  my $orig = \&ExtUtils::MakeMaker::WriteMakefile;
  my $writer = sub {
    my %params = @_;

    # Do nothing if not called from Makefile.PL
    my ($caller, $file, $line) = caller;
    (my $root = rel2abs($file)) =~ s/Makefile\.PL$//i or return;

    if (my $file = eval { Module::CPANfile->load(catfile($root, "cpanfile")) }) {
      my $prereqs = $file->prereqs;

      # Runtime requires => PREREQ_PM
      _merge(
        \%params,
        _get($prereqs, 'runtime', 'requires'),
        'PREREQ_PM',
      );

      # Build requires => BUILD_REQUIRES / PREREQ_PM
      _merge(
         \%params,
         _get($prereqs, 'build', 'requires'),
         _eumm('6.56') ? 'BUILD_REQUIRES' : 'PREREQ_PM',
      );

      # Test requires => TEST_REQUIRES / BUILD_REQUIRES / PREREQ_PM
      _merge(
         \%params,
         _get($prereqs, 'test', 'requires'),
         _eumm('6.63_03') ? 'TEST_REQUIRES' :
         _eumm('6.56') ? 'BUILD_REQUIRES' : 'PREREQ_PM',
      );

      # Configure requires => CONFIGURE_REQUIRES / ignored
      _merge(
         \%params,
         _get($prereqs, 'configure', 'requires'),
         _eumm('6.52') ? 'CONFIGURE_REQUIRES' : undef,
      );

      # Add myself to configure requires (if possible)
      _merge(
         \%params,
         {'ExtUtils::MakeMaker::CPANfile' => $VERSION},
         _eumm('6.52') ? 'CONFIGURE_REQUIRES' : undef,
      );

      # Set dynamic_config to 0 if not set explicitly
      if (!exists $params{META_ADD}{dynamic_config} &&
          !exists $params{META_MERGE}{dynamic_config}) {
          $params{META_MERGE}{dynamic_config} = 0;
      }

      # recommends, suggests, conflicts
      my $requires_2_0;
      for my $type (qw/recommends suggests conflicts/) {
          for my $phase (qw/configure build test runtime develop/) {
              my %tmp = %{$params{META_MERGE}{prereqs}{$phase} || {}};
              _merge(
                  \%tmp,
                  _get($prereqs, $phase, $type),
                  $type,
              );
              if ($tmp{$type}) {
                  $params{META_MERGE}{prereqs}{$phase} = \%tmp;
                  $requires_2_0 = 1;
              }
          }
      }
      if ($requires_2_0) { # for better recommends support
          $params{META_MERGE}{"meta-spec"} ||= {version => 2};
      }

      # XXX: better to use also META_MERGE when applicable?

      # As a small bonus, remove params that the installed version
      # of EUMM doesn't know, so that we can always write them
      # in Makefile.PL without caring about EUMM version.
      # (EUMM warns if it finds unknown parameters.)
      # As EUMM 6.17 is our prereq, we can safely ignore the keys
      # defined before 6.17.
      {
        last if _eumm('6.66_03');
        if (my $r = delete $params{TEST_REQUIRES}) {
          _merge(\%params, $r, 'BUILD_REQUIRES');
        }
        last if _eumm('6.56');
        if (my $r = delete $params{BUILD_REQUIRES}) {
          _merge(\%params, $r, 'PREREQ_PM');
        }

        last if _eumm('6.52');
        delete $params{CONFIGURE_REQUIRES};

        last if _eumm('6.47_01');
        delete $params{MIN_PERL_VERSION};

        last if _eumm('6.45_01');
        delete $params{META_ADD};
        delete $params{META_MERGE};

        last if _eumm('6.30_01');
        delete $params{LICENSE};
      }
    } else {
        print "cpanfile is not available: $@\n";
        exit 0; # N/A
    }

    $orig->(%params);
  };
  {
    no warnings 'redefine';
    *main::WriteMakefile =
    *ExtUtils::MakeMaker::WriteMakefile = $writer;
  }
}

sub _eumm {
  my $version = shift;
  eval { ExtUtils::MakeMaker->VERSION($version) } ? 1 : 0;
}

sub _get {
  my $prereqs = shift;
  eval { $prereqs->requirements_for(@_)->as_string_hash };
}

sub _merge {
  my ($params, $requires, $key) = @_;

  return unless $key;

  for (keys %{$requires || {}}) {
    my $version = _normalize_version($requires->{$_});
    next unless defined $version;

    if (not exists $params->{$key}{$_}) {
      $params->{$key}{$_} = $version;
    } else {
      my $prev = $params->{$key}{$_};
      if (version->parse($prev) < version->parse($version)) {
        $params->{$key}{$_} = $version;
      }
    }
  }
}

sub _normalize_version {
  my $version = shift;

  # shortcuts
  return unless defined $version;
  return $version unless $version =~ /\s/;

  # TODO: better range handling
  $version =~ s/(?:>=|==)\s*//;
  $version =~ s/,.+$//;

  return $version unless $version =~ /\s/;
  return;
}

1;

__END__

=encoding utf-8

=head1 NAME

ExtUtils::MakeMaker::CPANfile - cpanfile support for EUMM

=head1 SYNOPSIS

    # Makefile.PL
    use ExtUtils::MakeMaker::CPANfile;

    WriteMakefile(
      NAME => 'Foo::Bar',
      AUTHOR => 'A.U.Thor <author@cpan.org>',
    );

    # cpanfile
    requires 'ExtUtils::MakeMaker' => '6.17';
    on test => sub {
      requires 'Test::More' => '0.88';
    };

=head1 DESCRIPTION

ExtUtils::MakeMaker::CPANfile loads C<cpanfile> in your distribution
and modifies parameters for C<WriteMakefile> in your Makefile.PL.
Just use it instead of L<ExtUtils::MakeMaker> (which should be
loaded internally), and prepare C<cpanfile>.

As of version 0.03, ExtUtils::MakeMaker::CPANfile also removes
WriteMakefile parameters that the installed version of
ExtUtils::MakeMaker doesn't know, to avoid warnings.

=head1 LIMITATION

=head2 complex version ranges

As of this writing, complex version ranges are simply ignored.

=head2 dynamic config

Strictly speaking, C<cpanfile> is a Perl script, and may have some
conditions in it. That said, you don't need to run Makefile.PL
to determine prerequisites in most cases. Hence, as of 0.06,
ExtUtils::MakeMaker::CPANfile sets C<dynamic_config> to false
by default. If you do need a CPAN installer to run Makefile.PL
to customize prerequisites dynamically, set C<dynamic_config>
to true explicitly (via META_ADD/META_MERGE).

=head1 FOR MODULE AUTHORS

Though the minimum version requirement of ExtUtils::MakeMaker is
arbitrary set to 6.17 (the one bundled in Perl 5.8.1), you need
at least EUMM 6.52 (with CONFIGURE_REQUIRES support) when you
release a distribution.

=head1 LICENSE

Copyright (C) Kenichi Ishigaki.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Kenichi Ishigaki E<lt>ishigaki@cpan.orgE<gt>

=cut

